### DISCUSSION TITLE
Spinners shouldn't be able to be maxed out before the spinner object ends

### DISCUSSION THREAD

**Discussion by Sleepteiner** [ppy/osu#25213] [Ideas] [answered]
(This post is mostly copied from my changelog comment.)

I am not necessarily against making spinners easier to "complete", although there are plenty of arguments against doing that, but I am against having spinners be completable before the spinner object ends.

In the new spinner system on long spinners on low OD maps (something that isn't that uncommon), a large percentage of people will max out the long spinner multiple seconds before the spinner ends. This isn't fun and it messes with the intended flow of the map. The duration and end point of a spinner is an intentional and important decision that the mapper makes, therefore players should be encouraged to spin the whole way through a spinner.

How about instead of maxing out the spinner early, spinners could behave like in stable, but with a lower spinning RPM cap (the RPM number at the bottom should probably still go above the new cap to avoid confusion though). In stable, to max out a spinner, you don't have to just spin the equivalent amount of times that auto does, you have to maintain spinning above 477 RPM for the duration of the spinner.

In other words, I think a better system than maxing out spinners early would be to require players to maintain a spinning speed above the new completion requirement for the duration of the spinner. For example, on OD0 maps, players would have to keep their spinning speed above 250 RPM for the whole spinner duration to max out the spinner. They wouldn't be allowed to play half of it with 500 RPM and then just stop spinning. This would both make competition on spinners more accessible to new players and would respect the mapper's intent by encouraging players to continue playing spinners like mappers assumed most players would always play them when they were adding the spinner to their map.

(Also, if the max spinner score is based on OD, then that means that the difficulty adjust mod can be used to get more score even though it isn't currently meant to. If the max spinner score is to be based on OD going forward, then maybe it should be based on the original OD of the map, not the OD after mods affect the stat? Or probably better, maybe just base the max spinner score off of spinner duration alone and have OD only affect how fast you have to maintain spinning speed to get that same max spinner score?)

**Comment by marioman63** [comment]
I started playing again after taking a multi-month break, and noticed this quirk about spinners right away. I agree that it was fairly jarring to see, and it made spinners feel less rewarding even though technically not much has changed, except the rate at which bonus is accumulated based on my research on updates I missed over the last few months. 

It is incredibly noticeable on this [spinner test map ](https://osu.ppy.sh/beatmapsets/213663#osu/502237), which is a good demonstration of the issue described above: I tested the relatively recent changes to Lazer's spinners today and noticed that I completed the spinner before the halfway mark of the map. Compared to stable, where I can spin up to the end of the map and still accumulate score (I usually spin about 450 RPM by stable's calculations, so a little shy from the potential score cap in stable).

Before the recent changes to the spinner, I didn't even know there _was_ a score cap in lazer because it required most of the spinner to hit (unless you were a top player perhaps, then maybe there's a noticeable gap). Now that you can complete a spinner very early at a large range of skill levels, it makes them less satisfying in some cases. I like the proposed idea of rewarding stamina over burst spinning. It still maintains a sense of skill required to play the spinner, but like you say would make them more accessible and easier on the hand in some cases.

I think one of the concerns brought up and why the spinners seemed to be changed was because a lot of spinners do not leave ample time between their end and the next hit object, making it hard to "get out of spinner mode" and aim. While your proposal could potentially undo the fix to this concern, I'm not sure why it was a concern in the first place, since I was under the impression such mapping was considered "outdated" by the official ranking criteria, regardless of its popularity. As a result, I feel like this point is a non-issue.

**Comment by molneya** [comment]
Before this discussion or latest spinner changes were made I made a poll in a discord server asking some spinner players what they thought about adding an rpm cap. It was supported by quite a few players (see https://github.com/ppy/osu/issues/23732#issuecomment-1733753964 for poll context)

On the latest spinner changes, ppy and smoogi said they don't dislike the idea:
>There are also areas not addressed - peppy & I like the idea of capping the spinning RPM to enforce consistency, and this will be required for lazer's implementation. This will be done as a separate change. (from https://github.com/ppy/osu/pull/24932)

Hopefully this can get looked at one last time before leaderboards go live.

**Comment by Bubbleman532** [comment]
I feel like having an RPM cap on a spinner which exactly matches the required spin speed is too much, though I do recognise this is how it used to "work" before

I'm on board with capping the RPM as long as the spinner can still be maxed early, so maybe an RPM cap of 10-15% or something on top of the requirement.

Edit as a side thought: I feel maxing spinners is only really going to matter for competition once you get to 2x rate with rate changes being added, so however it works you'll still have a high maximum to reach for when looking for bonus score

**Reply by Sleepteiner** [reply]
Why is it "too much" to require players to spin for the whole duration of the spinner to get max score from that spinner?

My argument is not really made from the perspective of competition. It is from the perspective of the artistic preservation of mapping. If the game allows the player to max a spinner early, then that means that the game is punishing players for playing the whole spinner. If the game punishes players for playing an object in a complete way, then that messes with the flow of thousands of ranked maps.

If we want more leniency for players so that they don't actually have to spin the RPM cap the whole duration of the spinner, there could be a "wind up" time at the start of the spinner where the RPM cap starts very low and then increases to the intended RPM cap to give players enough time to get up to speed in a consistent way. Isn't that how it already works in stable? (EDIT: It looks like that "wind up" time functionality is already in lazer as well based on a little bit of testing. Correct me if I'm wrong though.)

**Comment by smoogipoo** [comment, marked as answer]
Tending to agree with the suggestion here, and I think peppy does as well. Implementation seems to be rather tricky.

**Reply by smoogipoo** [reply]
https://github.com/ppy/osu/issues/25805
